Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
+  }
+
+  public enum CurrentStatus {
+    Current = 0,
+    Obsolete = 1,
+    Verifying = 2
+  }
+
+  public enum LineVerificationStatus {
+    // Default value for every line, before the renderer figures it out.
+    Unknown = 0,
+    // For first-time computation not actively computing but soon. Synonym of "obsolete"
+    // (scheduledComputation)
+    Scheduled = 1,
+    // For first-time computations, actively computing
+    Verifying = 2,
+    VerifiedObsolete = 201,
+    VerifiedVerifying = 202,
+    // Also applicable for empty spaces if they are not surrounded by errors.
+    Verified = 200,
+    // For containers of other diagnostics nodes (e.g. methods)
+    ErrorRangeObsolete = 301,
+    ErrorRangeVerifying = 302,
+    ErrorRange = 300,
+    // For individual assertions in error ranges
+    ErrorRangeAssertionVerifiedObsolete = 351,
+    ErrorRangeAssertionVerifiedVerifying = 352,
+    ErrorRangeAssertionVerified = 350,
+    // For specific lines which have errors on it. They take over verified assertions
+    ErrorObsolete = 401,
+    ErrorVerifying = 402,
+    Error = 400,
+    // For lines containing resolution or parse errors
+    ResolutionError = 16
+  }
+
+  public record NodeDiagnostic(
+     // User-facing name
+     string DisplayName,
+     // Used to re-trigger the verification of some diagnostics.
+     string Identifier,
+     string Filename,
+     // The range of this node.
+     Range Range
+  ) {
+    // Overriden by checking children if there are some
+    public VerificationStatus StatusVerification { get; set; } = VerificationStatus.Unknown;
+
+    // Overriden by checking children if there are some
+    public CurrentStatus StatusCurrent { get; set; } = CurrentStatus.Obsolete;
+
+    // Used to relocate a node diagnostic and to determine which function is currently verifying
+    public Position Position => Range.Start;
+
+    /// Time and Resource diagnostics
+    public bool Started { get; set; } = false;
+    public bool Finished { get; set; } = false;
+    public DateTime StartTime { get; protected set; }
+    public DateTime EndTime { get; protected set; }
+    public int TimeSpent => (int)(Finished ? ((TimeSpan)(EndTime - StartTime)).TotalMilliseconds : Started ? (DateTime.Now - StartTime).TotalMilliseconds : 0);
+    // Resources allocated at the end of the computation.
+    public int ResourceCount { get; set; } = 0;
+
+
+
+    // If this node is an error, all the trace positions
+    public ImmutableList<Range> RelatedRanges { get; set; } = ImmutableList<Range>.Empty;
+
+    // Sub-diagnostics if any
+    public List<NodeDiagnostic> Children { get; set; } = new();
+    private List<NodeDiagnostic> NewChildren { get; set; } = new();
```

Comment:
To work with existing diagnostics and ensure that at the end everything is coherent,, I use several times a mechanism of "double buffering". I keep the previous (migrated) children, potentially updating them with the new verification results, but I also add new verification results in `NewChildren`, so that when the verification of this node finishes, I just replace `Children` with `NewChildren`.